File: tools/Word Document Editor/scripts/apply_change.py
#!/usr/bin/env python3
"""
Apply a specific track change in a Word document.

REQUIRES EXPLICIT USER CONFIRMATION - never applies changes automatically.

Usage:
    python3 apply_change.py document.docx --change-id 1
    python3 apply_change.py document.docx --change-id 1 --output new_document.docx
"""

import sys
import zipfile
import shutil
import argparse
from xml.etree import ElementTree as ET
from pathlib import Path
from docx_utils import list_track_changes, NS, get_text_from_element


def apply_change(docx_path, change_id, output_path=None, confirm=True):
    """
    Apply a specific track change.
    
    Args:
        docx_path: Path to input document
        change_id: ID of change to apply (from list_changes)
        output_path: Path to save output (None = overwrite original)
        confirm: Require user confirmation (default: True)
    """
    docx_path = Path(docx_path)
    
    if not docx_path.exists():
        print(f"❌ Error: File not found: {docx_path}", file=sys.stderr)
        sys.exit(1)
    
    # First, list changes to find the one we want
    changes = list_track_changes(docx_path)
    
    # Find the change by ID
    target_change = None
    for change in changes:
        if change['id'] == change_id:
            target_change = change
            break
    
    if not target_change:
        print(f"❌ Error: Change ID {change_id} not found", file=sys.stderr)
        print(f"   Run 'list_changes.py' to see available change IDs", file=sys.stderr)
        sys.exit(1)
    
    # Show change details
    print(f"\n📋 Change to apply:")
    print(f"   ID: {target_change['id']}")
    print(f"   Type: {target_change['type']}")
    print(f"   Author: {target_change['author']}")
    print(f"   Text: {repr(target_change['text'][:100])}")
    
    # Require explicit confirmation
    if confirm:
        print(f"\n⚠️  WARNING: This will apply the change to the document.")
        response = input("   Type 'APPLY' to confirm: ")
        if response != 'APPLY':
            print("❌ Cancelled - change not applied.")
            sys.exit(0)
    
    # Determine output path
    if output_path is None:
        output_path = docx_path
    else:
        output_path = Path(output_path)
    
    try:
        # Copy document to output location
        if output_path != docx_path:
            shutil.copy2(docx_path, output_path)
        
        # Open document
        with zipfile.ZipFile(output_path, 'a') as docx:
            # Read document.xml
            doc_xml = docx.read('word/document.xml')
            root = ET.fromstring(doc_xml)
            
            # Find and apply the change based on type
            applied = False
            
            if target_change['element'] == 'ins':
                # For insertions, we need to keep the content but remove the w:ins wrapper
                for ins in root.findall('.//w:ins', NS):
                    author = ins.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', '')
                    text = get_text_from_element(ins)
                    if author == target_change['author'] and text == target_change['text']:
                        # Move children to parent
                        parent = ins.getparent()
                        if parent is not None:
                            # Insert children before the ins element
                            for child in list(ins):
                                parent.insert(list(parent).index(ins), child)
                            parent.remove(ins)
                            applied = True
                            break
            
            elif target_change['element'] == 'del':
                # For deletions, remove the w:del element entirely
                for dele in root.findall('.//w:del', NS):
                    author = dele.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', '')
                    text = get_text_from_element(dele)
                    if author == target_change['author'] and text == target_change['text']:
                        parent = dele.getparent()
                        if parent is not None:
                            parent.remove(dele)
                            applied = True
                            break
            
            elif target_change['element'] in ['rPrChange', 'pPrChange']:
                # For formatting changes, remove the change wrapper but keep the formatting
                element_name = target_change['element']
                for fmt_change in root.findall(f'.//w:{element_name}', NS):
                    author = fmt_change.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', '')
                    if author == target_change['author']:
                        # Keep the formatting properties, remove the change wrapper
                        parent = fmt_change.getparent()
                        if parent is not None:
                            # Move formatting properties to parent
                            for child in list(fmt_change):
                                if child.tag.endswith('}rPr') or child.tag.endswith('}pPr'):
                                    # Replace parent's formatting with new formatting
                                    for existing in parent.findall('.//w:rPr', NS):
                                        parent.remove(existing)
                                    for existing in parent.findall('.//w:pPr', NS):
                                        parent.remove(existing)
                                    parent.append(child)
                            parent.remove(fmt_change)
                            applied = True
                            break
            
            if not applied:
                print(f"⚠️  Warning: Could not locate exact change element in XML", file=sys.stderr)
                print(f"   The change may have already been applied or document structure differs", file=sys.stderr)
                sys.exit(1)
            
            # Write back to document
            doc_xml_new = ET.tostring(root, encoding='unicode')
            # Word requires specific XML declaration
            doc_xml_new = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + doc_xml_new
            
            # Remove old file and write new one
            docx.writestr('word/document.xml', doc_xml_new.encode('utf-8'))
        
        print(f"\n✓ Success! Change applied and saved to: {output_path}")
        
    except Exception as e:
        print(f"❌ Error: {type(e).__name__}: {str(e)}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description='Apply a specific track change (requires explicit confirmation)'
    )
    parser.add_argument('document', help='Path to Word document (.docx)')
    parser.add_argument('--change-id', type=int, required=True,
                       help='ID of change to apply (from list_changes.py)')
    parser.add_argument('--output', '-o',
                       help='Output file path (default: overwrites original)')
    parser.add_argument('--no-confirm', action='store_true',
                       help='Skip confirmation prompt (NOT RECOMMENDED)')
    
    args = parser.parse_args()
    
    apply_change(
        args.document,
        args.change_id,
        output_path=args.output,
        confirm=not args.no_confirm
    )


if __name__ == '__main__':
    main()

File: tools/Word Document Editor/scripts/apply-change.js
#!/usr/bin/env node
/**
 * Apply (accept) a track change in a Word document.
 *
 * Usage:
 *     node apply-change.js document.docx --change-id 1
 */

import { readDocx, writeDocx, parseXml, serializeXml, select, listTrackChanges, getTextFromElement } from './docx-utils.js';
import { existsSync, copyFileSync } from 'fs';
import { createInterface } from 'readline';

/**
 * Prompt user for confirmation
 */
async function confirm(message) {
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(resolve => {
    rl.question(`${message} (y/N): `, answer => {
      rl.close();
      resolve(answer.toLowerCase() === 'y' || answer.toLowerCase() === 'yes');
    });
  });
}

/**
 * Apply a track change in a Word document
 */
async function applyChange(docxPath, changeId, options = {}) {
  const { outputPath = null, skipConfirmation = false } = options;

  // List changes first
  const changes = await listTrackChanges(docxPath);
  const change = changes.find(c => c.id === changeId);

  if (!change) {
    throw new Error(`Change ID ${changeId} not found. Run list-changes.js to see available changes.`);
  }

  // Show change details
  console.log('\nChange to apply:');
  console.log(`  Type: ${change.type}`);
  console.log(`  Author: ${change.author}`);
  console.log(`  Text: "${change.text.slice(0, 100)}${change.text.length > 100 ? '...' : ''}"`);

  // Confirm
  if (!skipConfirmation) {
    const confirmed = await confirm('\nApply this change?');
    if (!confirmed) {
      console.log('Operation cancelled.');
      process.exit(0);
    }
  }

  // Determine output path
  const finalOutputPath = outputPath || docxPath;

  // Copy file if output is different
  if (outputPath && outputPath !== docxPath) {
    copyFileSync(docxPath, outputPath);
  }

  // Read document
  const zip = await readDocx(outputPath || docxPath);

  // Read document.xml
  const docXml = await zip.file('word/document.xml')?.async('string');
  if (!docXml) {
    throw new Error('Invalid Word document structure');
  }
  const docDoc = parseXml(docXml);

  // Find the nth change element of the specified type
  const elementTag = change.element === 'ins' ? 'w:ins' : change.element === 'del' ? 'w:del' : `w:${change.element}`;
  const changeElements = select(`//${elementTag}`, docDoc);

  // Track which instance we're looking for based on order in original list
  let targetIndex = 0;
  for (let i = 0; i < changes.length; i++) {
    if (changes[i].id === changeId) break;
    if (changes[i].element === change.element) targetIndex++;
  }

  const targetElement = changeElements[targetIndex];
  if (!targetElement) {
    throw new Error('Could not locate change element in document');
  }

  const parent = targetElement.parentNode;

  if (change.type === 'Insertion') {
    // Accept insertion: replace w:ins with its children (keep the content)
    const children = Array.from(targetElement.childNodes);
    for (const child of children) {
      parent.insertBefore(child.cloneNode(true), targetElement);
    }
    parent.removeChild(targetElement);
  } else if (change.type === 'Deletion') {
    // Accept deletion: remove the w:del and its contents (content is deleted)
    parent.removeChild(targetElement);
  } else if (change.type.startsWith('Formatting')) {
    // Accept formatting: remove the *PrChange element (keep new formatting)
    parent.removeChild(targetElement);
  }

  // Write back document.xml
  zip.file('word/document.xml', serializeXml(docDoc));

  // Save
  await writeDocx(zip, finalOutputPath);

  return { changeId, type: change.type };
}

// CLI
function showHelp() {
  console.log(`
Apply (accept) a track change in a Word document

Usage:
  node apply-change.js <document> --change-id <N> [options]

Arguments:
  document          Path to Word document (.docx)

Options:
  --change-id N     ID of change to apply (from list-changes output)
  --output, -o      Output file path (default: overwrites original)
  --yes, -y         Skip confirmation prompt
  --help, -h        Show this help message

Examples:
  node apply-change.js document.docx --change-id 1
  node apply-change.js document.docx --change-id 2 --output modified.docx
`);
  process.exit(0);
}

const args = process.argv.slice(2);

if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
  showHelp();
}

const docxPath = args[0];
let changeId = null;
let outputPath = null;
let skipConfirmation = false;

for (let i = 1; i < args.length; i++) {
  if (args[i] === '--change-id' && args[i + 1]) {
    changeId = parseInt(args[++i], 10);
  } else if ((args[i] === '--output' || args[i] === '-o') && args[i + 1]) {
    outputPath = args[++i];
  } else if (args[i] === '--yes' || args[i] === '-y') {
    skipConfirmation = true;
  }
}

if (changeId === null) {
  console.error('Error: --change-id is required');
  process.exit(1);
}

if (!existsSync(docxPath)) {
  console.error(`Error: File not found: ${docxPath}`);
  process.exit(1);
}

try {
  const result = await applyChange(docxPath, changeId, { outputPath, skipConfirmation });
  console.log(`\nSuccess! Change ${result.changeId} (${result.type}) applied.`);
  console.log(`Saved to: ${outputPath || docxPath}`);
} catch (e) {
  console.error(`Error: ${e.message}`);
  process.exit(1);
}

